app/lib/session_encryptor.rb

```diff
@@ -0,0 +1,59 @@
+class SessionEncryptor
+  KEY_LEN = ActiveSupport::MessageEncryptor.key_len
+
+  def initialize(session_signing_key:)
+    @session_signing_key = session_signing_key
+  end
+
+  def encrypt_session(access_token:, refresh_token:)
+    encrypt_string(
+      {
+        access_token: Base64.encode64(access_token),
+        refresh_token: Base64.encode64(refresh_token),
+      }.to_json,
+    )
+  end
+
+  def decrypt_session(ciphertext)
+    return if ciphertext.blank?
+
+    plaintext = decrypt_string(ciphertext)
+    return unless plaintext
+
+    hash = JSON.parse(plaintext)
+
+    {
+      access_token: Base64.decode64(hash["access_token"]),
+      refresh_token: Base64.decode64(hash["refresh_token"]),
+    }
+  end
+
+protected
+
+  attr_reader :session_signing_key
+
+  def encrypt_string(plaintext)
+    salt = SecureRandom.hex KEY_LEN
+    ciphertext = encryptor(salt).encrypt_and_sign(plaintext)
+    Base64.urlsafe_encode64("#{salt}$$#{ciphertext}")
```

Comment:
Why do we want to encode the salt too? Isn't that only needed to create the ciphertext?

Reply:
The salt is also needed for decryption, so it needs to be stored in the user's session too.

Reply:
Never mind, I see that it's used when decrypting.

Reply:
We wrote at the same time 😅